Cleanup library headers, resolve dependency cycles

BenWiederhake opened the pull request:
This PR makes sure that each library header can be "compiled" stand-alone. The motivation was originally to find headers that have missing includes and simply don't compile correctly, and only work by pure luck. In particular, I think it's a good idea to make sure that the following never fails to compile:

```
#include <LibFoo/Bar.h>
// Nothing else in this .cpp file!
```

I used a tool to compile headers like this, but all the changes are manual. As such, I am reasonably sure that I didn't make any silly mistakes as in #10001 (famous last words).

Specifically:
- (About 35 commits) Some of the commits simply add missing headers. These "missing" headers never caused problems so far, because all include-sites so far have happened to include the "missing" headers already. Therefore, neither build time nor correctness nor performance should be noticeably affected.
- (1 commit) Add missing forward-declaration: Obviously, neither build time nor correctness nor performance should be affected.
- (1 commit) Clean up static function in header: This should decrease object size by a handful of bytes, increase running time by a handful of instructions (or not at all) for a function that isn't performance-sensitive, but makes compilation saner.
- (1 commit) Fix redeclaration: I honestly don't know why this code was allowed by the compiler. This commit shouldn't affect either build time or performance or memory layout, but assigns saner identifiers.
- (3 commits) Resolve cyclic dependency: These might be the most controversial; see commit descriptions for details. The theme is always the same: File A.h includes file B.h, and file B.h fails to compile on its own because it assumes that file A.h is already being included. Some commits might affect performance, but I wouldn't be convinced by it.

The tool I wrote to detect these issues isn't ready for inclusion yet, but see #FIXME for a draft version. Since the tool will have a separate host of issues, and these changes still make sense in isolation, and reviewing this PR is going to be quite some work anyway, I want to make the "tool PR" separate from this one.

linusg commented on Userland/Libraries/LibJS/Runtime/Temporal/PlainDate.h:
```diff
@@ -7,8 +7,10 @@
 
 #pragma once
 
+// FIXME
```
?

linusg commented on Userland/Libraries/LibJS/Runtime/Temporal/PlainDate.h:
```diff
@@ -7,8 +7,10 @@
+// FIXME
 #include <LibJS/Runtime/Completion.h>
 #include <LibJS/Runtime/Object.h>
+#include <LibJS/Runtime/Value.h>
```
You want to add a header for Value (implicitly pulled in literally everywhere) but not for `GlobalObject`, `FunctionObject`, `String`, and `StringView` (all used in this header but not explicitly included)?

BenWiederhake replied:
To answer both of your comments: Originally, it didn't compile without the #include for Value.h, so I added it in. Then, one day when rebasing, I ran into a merge conflict, wanted to make it compile again so I accepted *all* the includes with a FIXME to remind myself to pare the includes down again. Well, apparently I forgot this one, sorry! :sweat_smile:

BenWiederhake commented:
Changes since first push:
- Remove accident due to a merge request (thanks Linus!)
- Fix two commit subjects ("BinaryHeap" → "BinaryHeap.h" and "LibWasm: Add missing forward declaration to Printer/Printer.h")
- Remove now-obsolete `#include <LibGUI/Forward.h>` from "LibSyntax: Add missing headers to HighlighterClient.h"

CI looks like it detected additional problems, I'll investigate that soon.

BenWiederhake commented:
Changes since last push:
- In `LibCompress/DeflateTables.h`, include `<stdint.h>` instead of `<LibC/stdint.h>`. Of course the second one isn't available while building lagom.

BenWiederhake commented:
CI failure looks a lot like [regular flakiness](https://dev.azure.com/SerenityOS/SerenityOS/_build/results?buildId=2902&view=logs&jobId=42f0633d-9ac7-51d1-0d17-9c28359eb2e8&j=42f0633d-9ac7-51d1-0d17-9c28359eb2e8&t=13456976-1d69-5c50-e408-a62254233cc1).

Changes since last push:
- Rebase for convenience
- *and* enjoy the free side-effect of a CI re-run, to check whether CI detected an actual problem or was just flaky.

BenWiederhake commented:
I don't know what's wrong with CI :question:

BenWiederhake commented:
Changes since last push: Rebased, and demonstrate that CI is just too flaky.
